src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp

```diff
@@ -362,28 +368,110 @@ void GenericOTARequestorDriver::DefaultProviderTimerHandler(System::Layer * syst
     SendQueryImage();
 }
 
-void GenericOTARequestorDriver::StartDefaultProviderTimer()
+void GenericOTARequestorDriver::StartPeriodicQueryTimer()
 {
-    ChipLogProgress(SoftwareUpdate, "Starting the Default Provider timer, timeout: %u seconds",
+    ChipLogProgress(SoftwareUpdate, "Starting the periodic query timer, timeout: %u seconds",
                     (unsigned int) mPeriodicQueryTimeInterval);
     ScheduleDelayedAction(
         System::Clock::Seconds32(mPeriodicQueryTimeInterval),
         [](System::Layer *, void * context) {
-            (static_cast<GenericOTARequestorDriver *>(context))->DefaultProviderTimerHandler(nullptr, context);
+            (static_cast<GenericOTARequestorDriver *>(context))->PeriodicQueryTimerHandler(nullptr, context);
         },
         this);
 }
 
-void GenericOTARequestorDriver::StopDefaultProviderTimer()
+void GenericOTARequestorDriver::StopPeriodicQueryTimer()
 {
-    ChipLogProgress(SoftwareUpdate, "Stopping the Default Provider timer");
+    ChipLogProgress(SoftwareUpdate, "Stopping the Periodic Query timer");
     CancelDelayedAction(
         [](System::Layer *, void * context) {
-            (static_cast<GenericOTARequestorDriver *>(context))->DefaultProviderTimerHandler(nullptr, context);
+            (static_cast<GenericOTARequestorDriver *>(context))->PeriodicQueryTimerHandler(nullptr, context);
+        },
+        this);
+}
+
+void GenericOTARequestorDriver::WatchdogTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    ChipLogProgress(SoftwareUpdate, "Watchdog timer handler is invoked");
+
+    OTAUpdateStateEnum currentState = mRequestor->GetCurrentUpdateState();
+
+    switch (currentState)
+    {
+    case OTAUpdateStateEnum::kIdle:
+    case OTAUpdateStateEnum::kUnknown:
+        // OTA Requestor is not stuck in non-idle state.  Restart watchdog timer.
+        StartWatchdogTimer();
+        break;
+    case OTAUpdateStateEnum::kQuerying:
+    case OTAUpdateStateEnum::kDelayedOnQuery:
+    case OTAUpdateStateEnum::kDownloading:
+    case OTAUpdateStateEnum::kApplying:
+    case OTAUpdateStateEnum::kDelayedOnApply:
+        UpdateDiscontinued();
+        mRequestor->CancelImageUpdate();
+        mRequestor->Reset();
+        StartPeriodicQueryTimer();
+        break;
+    case OTAUpdateStateEnum::kRollingBack:
+    case OTAUpdateStateEnum::kDelayedOnUserConsent:
```

Comment:
I don't really see a need to differentiate all the states here. Specifically, if you were in `kQuerying` state when the watchdog timer expired, doesn't really make a whole lot of sense to call `CancelImageUpdate` but we can do it anyway since it won't hurt. If going on that, we could just perform the same "revert to normal" operations for every case.

Reply:
I wanted to make sure every case was covered so listed them all.  I've moved kQuerying and kDelayedOnQuery down to just do the Reset() instead of also doing CancelImageUpdate() now.

Reply:
We could probably simplify logic and just do whenever watch dog timer expired.. which we hope is never :) 
```
         UpdateDiscontinued();
         mRequestor->CancelImageUpdate();
         StartPeriodicQueryTimer();
```

Reply:
Will make this change.